src/object.d

```diff
@@ -3127,6 +3205,164 @@ inout(V) get(K, V)(inout(V[K])* aa, K key, lazy inout(V) defaultValue)
     }
 }
 
+unittest
+{
+    static class T
+    {
+        static size_t count;
+        this() { ++count; }
+    }
+
+    T[string] aa;
+
+    auto a = new T;
+    aa["foo"] = a;
+    assert(T.count == 1);
+    auto b = aa.get("foo", new T);
+    assert(T.count == 1);
+    assert(b is a);
+    auto c = aa.get("bar", new T);
+    assert(T.count == 2);
+    assert(c !is a);
+
+    //Obviously get doesn't add.
+    assert("bar" !in aa);
+}
+
+/***********************************
+ * Looks up key; if it exists returns corresponding value else evaluates
+ * value, adds it to the associative array and returns it.
+ * Params:
+ *      aa =     The associative array.
+ *      key =    The key.
+ *      value =  The required value.
+ * Returns:
+ *      The value.
+ */
+ref V require(K, V)(ref V[K] aa, K key, lazy V value = V.init) pure
```

Comment:
I would remove the `pure` annotation and make the template works its magic, or is there a specific reason you put it ?

Reply:
@Geod24 I Just copied the function attributes from the `get` function.

Reply:
@Geod24 Just so I am clear this is the 'magic' you are referring to: http://klickverbot.at/blog/2012/05/purity-in-d/#templates-and-purity i.e templates infer pure?
Presumably this wasn't the case when 'get' was implemented.

Reply:
Template infers attributes, so `@nogc`, `@safe`/`@system`, and `pure`
In this case since only `pure` and `nothrow` are put on the `extern (C)` declaration, that's as much as could get infered. Maybe there were some inference bug at the time ?

You can make sure this works by making a `pure unittest`